lib/src/bridge/characteristics_mixin.dart

```diff
@@ -11,23 +11,23 @@ mixin CharacteristicsMixin on FlutterBLE {
     String transactionId,
   ) =>
       _methodChannel
-          .invokeMethod(
+          .invokeMethod<String>(
             MethodName.readCharacteristicForIdentifier,
             <String, dynamic>{
               ArgumentName.characteristicIdentifier: characteristicIdentifier,
               ArgumentName.transactionId: transactionId
             },
           )
-          .catchError((errorJson) =>
-              Future.error(BleError.fromJson(jsonDecode(errorJson.details))))
+          .catchError((errorJson) async =>
+            throw BleError.fromJson(jsonDecode(errorJson.details))
+          )
           .then((rawValue) {
-            String rawJsonValue = "";
-            if (rawValue is String) {
-              rawJsonValue = rawValue;
+            if (rawValue == null) {
```

Comment:
That's fine. I understand, you see the correct semantics for this domain much better. It's just that, there were some -- and correct me if I'm wrong where a function returning a future, could become into a hanging state and never actually resolve, only on timeout if the caller sets one. Or is it always guaranteed from the native side that 'some' `.then()` (or `error()`) will always be executed?

The reason I don't get this is that there is this `withResponse` flag user can pass in and in the case of that being false, how will the returned future behave??
To me, it looks like it will never resolve without a(n unnecessary) timeout in the case of `withResponse = false`, and even then it wouldn't return a characteristic in the future raising a _TypeError in new typesafer dart runtime.

and on `withResponse = true` it behaves as expected ofc

This is why I thought it would make sense to make it return a nullable resolving future, that is guaranteed to always resolve, making the caller code much cleaner.

To sum up: 
- I will revert the commit, but pls consider these concerns, as these are changes better be done on a major version bump, as it introduces minor API amendments, potentially breaking adopters code also.
- What does Future<Characteristic> resolve to exactly on `withResponse = false`?

thanks

Reply:
Response is a byte array from the server (peripheral). If the central (client) does not await for the response, the operation is finished immediately after sending the write request.

Native plugin always returns either characteristic or error. The whole characteristic is returned regardless of the response for convenience - this way you can write to a characteristic from peripheral, save the reference to the characteristic and continue operations on it.